roxie/roxiemem/roxiemem.cpp

```diff
@@ -3712,7 +3723,8 @@ char * ChunkedHeaplet::allocateSingle(unsigned allocated, bool incCounter, unsig
 
             CChunkedHeap * chunkHeap = static_cast<CChunkedHeap *>(heap);
             unsigned maxAllocs = chunkHeap->maxChunksPerPage();
-            if (numAllocs == maxAllocs)
+            unsigned minFree = chunkHeap->minScanFree(); // Scanning when there are only a few spare slots becomes pathological, so give up early.
+            if (numAllocs + minFree > maxAllocs)
```

Comment:
ok.  I guess I was also thinking if minFree was < some percentage of maxAllocs then skip.